File: Pharmaceutical_Company_Example/03_Behavioral/Java/Strategy.java
// Strategy Interface
interface RegulatoryStrategy {
    void approveApplication(String drugName);
}

// Concrete Strategy: Standard Approval
class StandardApprovalStrategy implements RegulatoryStrategy {
    @Override
    public void approveApplication(String drugName) {
        System.out.println("Applying Standard Approval Process for drug: " + drugName);
        System.out.println(drugName + " has been approved using the Standard Approval Strategy.");
    }
}

// Concrete Strategy: Accelerated Approval
class AcceleratedApprovalStrategy implements RegulatoryStrategy {
    @Override
    public void approveApplication(String drugName) {
        System.out.println("Applying Accelerated Approval Process for drug: " + drugName);
        System.out.println(drugName + " has been approved using the Accelerated Approval Strategy.");
    }
}

// Context
class DrugApprovalContext {
    private RegulatoryStrategy strategy;

    public DrugApprovalContext(RegulatoryStrategy strategy) {
        this.strategy = strategy;
    }

    public void setStrategy(RegulatoryStrategy newStrategy) {
        this.strategy = newStrategy;
    }

    public void approveDrug(String drugName) {
        strategy.approveApplication(drugName);
    }
}

// Main class to test the Strategy pattern
public class Main {
    public static void main(String[] args) {
        // Create Context with Standard Approval Strategy
        DrugApprovalContext approvalContext = new DrugApprovalContext(new StandardApprovalStrategy());

        // Approve drug using Standard Approval Strategy
        approvalContext.approveDrug("Test Small Molecule 01");

        // Switch to Accelerated Approval Strategy
        approvalContext.setStrategy(new AcceleratedApprovalStrategy());
        approvalContext.approveDrug("Test Small Molecule 02");
    }
}

File: Pharmaceutical_Company_Example/03_Behavioral/Python/Strategy.py
from abc import ABC, abstractmethod

# Define the Strategy Interface
class RegulatoryStrategy(ABC):
    @abstractmethod
    def approve_application(self, drug_name):
        """This method should be overridden by concrete strategies!"""
        pass

# Concrete Strategy: Standard Approval
class StandardApprovalStrategy(RegulatoryStrategy):
    def approve_application(self, drug_name):
        print(f"Applying Standard Approval Process for drug: {drug_name}")
        print(f"{drug_name} has been approved using the Standard Approval Strategy.")

# Concrete Strategy: Accelerated Approval
class AcceleratedApprovalStrategy(RegulatoryStrategy):
    def approve_application(self, drug_name):
        print(f"Applying Accelerated Approval Process for drug: {drug_name}")
        print(f"{drug_name} has been approved using the Accelerated Approval Strategy.")

# Context
class DrugApprovalContext:
    def __init__(self, strategy: RegulatoryStrategy):
        self._strategy = strategy

    def set_strategy(self, new_strategy: RegulatoryStrategy):
        self._strategy = new_strategy

    def approve_drug(self, drug_name):
        self._strategy.approve_application(drug_name)

# Example Usage
# Create Context
approval_context = DrugApprovalContext(StandardApprovalStrategy())

# Set Standard Approval Strategy
approval_context.approve_drug("Test Small Molecule 01")

# Set Accelerated Approval Strategy
approval_context.set_strategy(AcceleratedApprovalStrategy())
approval_context.approve_drug("Test Small Molecule 02")
